File: Corte_II/01-Anotaciones/Ordenamiento/Sorts.py
# Comparo dos vecinos si estan al reves los intercambio y repito el proceso 
# hasta que nadie mas se mueva bubble sort

def bubble_sort(arr):
    n = len(arr)
    for i in range(n):
        swapped = False
        for j in range(0, n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
        if not swapped:
            break
    return arr



# Buscar el más pequeño y ponerlo al 
# principio selection sort

def selection_sort(arr):
    n = len(arr)
    for i in range(n):
        min_idx = i
        for j in range(i + 1, n):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
    return arr


# Tomo el segubndo y lo inserto donde va 
# respecto al primero insertion sort

def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    return arr


#Pruebas de sort


datos = [64, 25, 12, 22, 11, 90, 45, 33]

resultado1 = bubble_sort(datos)
print(f"Resultado por bubble sort: {resultado1}")

resultado2 = selection_sort(datos)
print(f"Resultado por selection sort: {resultado2}")

resultado3 =  insertion_sort(datos)
print(f"Resultado por insertion sort: {resultado3}")

#-----------------------------------------------------------------------------------------------------------------


def bubble_sort_comparaciones_e_intercambios(arr):
    n = len(arr)
    comparaciones = 0
    intercambios = 0
    
    for i in range(n):
        swapped = False
        for j in range(0, n - i - 1):
            comparaciones += 1  # Se realiza la comparación
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                intercambios += 1  # Se realiza el intercambio
                swapped = True
        if not swapped:
            break
            
    return arr, comparaciones, intercambios


def selection_sort_comparaciones_e_intercambios(arr):
    n = len(arr)
    comparaciones = 0
    intercambios = 0
    
    for i in range(n):
        min_idx = i
        for j in range(i + 1, n):
            comparaciones += 1  # Se compara para hallar el mínimo
            if arr[j] < arr[min_idx]:
                min_idx = j
                
        # Solo intercambia si encontró un elemento menor en una posición distinta
        if min_idx != i:
            arr[i], arr[min_idx] = arr[min_idx], arr[i]
            intercambios += 1
            
    return arr, comparaciones, intercambios


def insertion_sort_comparaciones_e_intercambios(arr):
    comparaciones = 0
    intercambios = 0  # En insertion sort representa los desplazamientos/asignaciones
    
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        
        while j >= 0:
            comparaciones += 1  # Evaluación de la condición (arr[j] > key)
            if arr[j] > key:
                arr[j + 1] = arr[j]
                intercambios += 1  # Desplazamiento del elemento hacia la derecha
                j -= 1
            else:
                break
                
        arr[j + 1] = key
        
    return arr, comparaciones, intercambios


# Pruebas parte 2
#Se usa .copy() para no alterar datos, pues si se usa la misma lista todos la modificarian asi que .copy() es como "clonar" datos 

datos = [64, 25, 12, 22, 11, 90, 45, 33]

# 1. Bubble Sort (usamos .copy() para no alterar 'datos')
res1, comp1, inter1 = bubble_sort_comparaciones_e_intercambios(datos.copy())
print(f"\nBubble Sort con comparaciones e intercambios = Resultado: {res1} | Comparaciones: {comp1} | Intercambios: {inter1}")

# 2. Selection Sort
res2, comp2, inter2 = selection_sort_comparaciones_e_intercambios(datos.copy())
print(f"Selection Sort con comparaciones e intercambios = Resultado: {res2} | Comparaciones: {comp2} | Intercambios: {inter2}")

# 3. Insertion Sort
res3, comp3, inter3 = insertion_sort_comparaciones_e_intercambios(datos.copy())
print(f"Insertion Sort con comparaciones e intercambios = Resultado: {res3} | Comparaciones: {comp3} | Desplazamientos: {inter3}")

File: Corte_II/01-Anotaciones/Ordenamiento/Sorts.cpp
#include <iostream>

using namespace std;

/* Bubble Sort */
void bubbleSort(int arr[], int n, int &comparaciones, int &intercambios) {
    comparaciones = 0;
    intercambios = 0;
    
    for (int i = 0; i < n - 1; i++) {
        bool swapped = false;
        for (int j = 0; j < n - i - 1; j++) {
            comparaciones++; // Se realiza la comparación
            if (arr[j] > arr[j + 1]) {
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
                intercambios++; // Se realiza el intercambio
                swapped = true;
            }
        }
        if (!swapped) break; // Optimización si ya está ordenado
    }
}

/* Selection Sort */
void selectionSort(int arr[], int n, int &comparaciones, int &intercambios) {
    comparaciones = 0;
    intercambios = 0;

    for (int i = 0; i < n - 1; i++) {
        int minIdx = i;
        for (int j = i + 1; j < n; j++) {
            comparaciones++; // Se realiza la comparación para hallar el mínimo
            if (arr[j] < arr[minIdx]) {
                minIdx = j;
            }
        }
        // Solo cuenta el intercambio si realmente cambia de posición
        if (minIdx != i) {
            int temp = arr[minIdx];
            arr[minIdx] = arr[i];
            arr[i] = temp;
            intercambios++;
        }
    }
}

/* Insertion Sort */
void insertionSort(int arr[], int n, int &comparaciones, int &desplazamientos) {
    comparaciones = 0;
    desplazamientos = 0; // Representa los desplazamientos de elementos

    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;

        while (j >= 0) {
            comparaciones++; // Se evalúa la condición de comparación
            if (arr[j] > key) {
                arr[j + 1] = arr[j];
                desplazamientos++; // Se desplaza el elemento a la derecha
                j--;
            } else {
                break;
            }
        }
        arr[j + 1] = key;
    }
}

// Función auxiliar para copiar arreglos nativos
void copiarArreglo(const int origen[], int destino[], int n) {
    for (int i = 0; i < n; i++) {
        destino[i] = origen[i];
    }
}

// Función auxiliar para imprimir el arreglo
void imprimirArreglo(const int arr[], int n) {
    cout << "[";
    for (int i = 0; i < n; i++) {
        cout << arr[i] << (i < n - 1 ? ", " : "");
    }
    cout << "]";
}

int main() {
    int datosOriginales[] = {64, 25, 12, 22, 11, 90, 45, 33};
    int n = sizeof(datosOriginales) / sizeof(datosOriginales[0]);
    
    // Arreglo donde crearemos las copias para cada prueba
    int datosCopia[n];
    int comp, inter;

    // 1. Bubble Sort
    copiarArreglo(datosOriginales, datosCopia, n); // Creamos el "clon"
    bubbleSort(datosCopia, n, comp, inter);
    cout << "Bubble Sort  ---> Resultado: ";
    imprimirArreglo(datosCopia, n);
    cout << " | Comparaciones: " << comp << " | Intercambios: " << inter << endl;

    // 2. Selection Sort
    copiarArreglo(datosOriginales, datosCopia, n); // Creamos el "clon"
    selectionSort(datosCopia, n, comp, inter);
    cout << "Selection Sort -----> Resultado: ";
    imprimirArreglo(datosCopia, n);
    cout << " | Comparaciones: " << comp << " | Intercambios: " << inter << endl;

    // 3. Insertion Sort
    copiarArreglo(datosOriginales, datosCopia, n); // Creamos el "clon"
    insertionSort(datosCopia, n, comp, inter);
    cout << "Insertion Sort ----> Resultado: ";
    imprimirArreglo(datosCopia, n);
    cout << " | Comparaciones: " << comp << " | Desplazamientos: " << inter << endl;

    return 0;
}